Retrieve all device states through 1 request

ehendrix23 opened the pull request:
Retrieving device states (open, close) from MyQ are now done through 1 request and stored internally. Through each device an update can still be requested however retrieval from the MyQ cloud is restricted to once every 5 seconds.

Requests to the MyQ cloud will also be sent synchronously ensuring only 1 request can be sent at a time.

These changes are to minimize requests to MyQ as there definitely seems to be a throttler on their end for how many requests can come in per account or so.

Note, just put this together today. Working here but would be good if we can get some more people involved to test these updates.

ehendrix23 committed:
Use 1 request for all devices

Changes to use 1 request to get state for all devices instead of individual requests. This to reduce number of calls to MyQ as there seems to be some kind of limiter on there.

ehendrix23 committed:
Include exception class to error

Include the exception class to error to log if issue is timeout or other error.

ehendrix23 committed:
Added missing typing

-) Added missing typing to changed/added methods
-) Moved MyQDevice import to method where it is required.

ehendrix23 committed:
Slowly increasing timeout

Starting timeout a lot smaller and slowly increasing on each timeout instead.

ehendrix23 committed:
Remove lock on request

Reverted the lock for request back, issue with this is that it might block request for open/close if getting status is timing out.

ehendrix23 committed:
Request new security token if expired

Request a new security token if it is expired.
Also new instance parameter self.online that is set True if able to retrieve states from MyQ and False if not.

ehendrix23 committed:
Set unavailable if unable to retrieve states from cloud

Set available property to false if unable to retrieve states from MyQ cloud.
Some lint issues.

ehendrix23 commented:
@arraylabs , been running with this updated version now for about almost a month, with the last change since 1/17. Have had no issues and my refresh time right now is set to 5 seconds. Sometimes there is a timeout, very rarely it happens to be that I get the warning it is taking longer then 10 seconds to update.

I believe this PR is ready to be merged so that we can then update HA to the newer version of pymyq.

towerhand commented:
I haven’t been able to test this PR, but I do get a lot of timeouts and update failures without it. I hope this will fix it.

ehendrix23 committed:
Added easy way to dump information retrieved

Added an easier method to dump the JSON info received from MyQ using the example.py.

arraylabs commented:
@ehendrix23 thanks for the work, this will make alot of people rest a little easier and calm down their logs!

arraylabs commented:
@ehendrix23 it has been uploaded to pypi.

ehendrix23 commented:
@arraylabs, Thanks. I'll open a PR now to update version requirement for MyQ in HA.
